src/librustc_data_structures/logged_unification_table.rs

```diff
@@ -0,0 +1,204 @@
+use rustc_index::vec::Idx;
+
+use crate::modified_set as ms;
+use crate::snapshot_vec as sv;
+use crate::unify as ut;
+use crate::unify_log as ul;
+
+use ena::undo_log::{Rollback, UndoLogs};
+
+pub enum UndoLog<K: ut::UnifyKey, I = K> {
+    Relation(sv::UndoLog<ut::Delegate<K>>),
+    UnifyLog(ul::Undo<I>),
+    ModifiedSet(ms::Undo<I>),
+}
+
+impl<K: ut::UnifyKey, I> From<sv::UndoLog<ut::Delegate<K>>> for UndoLog<K, I> {
+    fn from(l: sv::UndoLog<ut::Delegate<K>>) -> Self {
+        UndoLog::Relation(l)
+    }
+}
+
+impl<K: ut::UnifyKey, I> From<ul::Undo<I>> for UndoLog<K, I> {
+    fn from(l: ul::Undo<I>) -> Self {
+        UndoLog::UnifyLog(l)
+    }
+}
+
+impl<K: ut::UnifyKey, I> From<ms::Undo<I>> for UndoLog<K, I> {
+    fn from(l: ms::Undo<I>) -> Self {
+        UndoLog::ModifiedSet(l)
+    }
+}
+
+impl<K: ut::UnifyKey, I: Idx> Rollback<UndoLog<K, I>> for LoggedUnificationTableStorage<K, I> {
+    fn reverse(&mut self, undo: UndoLog<K, I>) {
+        match undo {
+            UndoLog::Relation(undo) => self.relations.reverse(undo),
+            UndoLog::UnifyLog(undo) => self.unify_log.reverse(undo),
+            UndoLog::ModifiedSet(undo) => self.modified_set.reverse(undo),
+        }
+    }
+}
+
+pub struct LoggedUnificationTableStorage<K: ut::UnifyKey, I: Idx = K> {
+    relations: ut::UnificationTableStorage<K>,
+    unify_log: ul::UnifyLog<I>,
+    modified_set: ms::ModifiedSet<I>,
+}
+
+pub struct LoggedUnificationTable<'a, K: ut::UnifyKey, I: Idx, L> {
+    storage: &'a mut LoggedUnificationTableStorage<K, I>,
+    undo_log: L,
+}
+
+impl<K, I> LoggedUnificationTableStorage<K, I>
+where
+    K: ut::UnifyKey + From<I>,
+    I: Idx + From<K>,
+{
+    pub fn new() -> Self {
+        Self {
+            relations: Default::default(),
+            unify_log: ul::UnifyLog::new(),
+            modified_set: ms::ModifiedSet::new(),
+        }
+    }
+
+    pub fn with_log<L>(&mut self, undo_log: L) -> LoggedUnificationTable<'_, K, I, L> {
```

Comment:
The name of `LoggedUnificationTable` is maybe suboptimal? It seems to overlap with the `with_log` method here somehow to me (where you are providing the log)